syntax = "proto3";
package gen;

import "Address.proto";

enum MessageType {
    PING = 0;
    ACK = 1;
    PING_REQ = 2;
    JOINREQ = 3;
    JOINREP = 4;
}

enum GossipEventTypes {
    JOINED = 0;
    FAILED = 1;
}

message Gossip {
    GossipEventTypes event = 1;
    Address address = 2;
    string id = 3;
}

message PingFields {
    Address targetAddress = 1;
}

message PingReqFields {
    Address targetAddress = 1;
}

message AckFields {
    string originalId = 1;
}

message Message {
    MessageType messageType = 1;
    Address sourceAddress = 2;
    Address destinationAddress = 3;
    string id = 4;

    oneof customFields {
        PingReqFields pingReqFields = 5;
        AckFields ackFields = 6;
        PingFields pingFields = 7;
    }

    repeated Gossip gossips = 8;
}